profiler/shakesapp/shakesapp/server.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package shakesapp defines a server which can be queried to determined
+// many times a string appears in the works of Shakespeare, and a client
+// which can be used to send load to that server.
+package shakesapp
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"regexp"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+// NewServer returns an implementation of the server for ShakespeareService
+// (defined in shakesapp.proto).
+func NewServer() ShakespeareServiceServer {
+	return &server{}
+}
+
+type server struct{}
+
+const bucketName = "dataflow-samples"
+const bucketPrefix = "shakespeare/"
+
+// GetMatchCount implements a server for ShakespeareService.
+func (s *server) GetMatchCount(ctx context.Context, req *ShakespeareRequest) (*ShakespeareResponse, error) {
+	resp := &ShakespeareResponse{}
+	texts, err := readFiles(ctx, bucketName, bucketPrefix)
+	if err != nil {
+		return resp, fmt.Errorf("fails to read files: %s", err)
+	}
+	for _, text := range texts {
+		for _, line := range strings.Split(text, "\n") {
+			line, query := strings.ToLower(line), strings.ToLower(req.Query)
+			// TODO: Compiling and matching a regular expression on every request
+			// might be too expensive? Consider optimizing.
+			isMatch, err := regexp.MatchString(query, line)
+			if err != nil {
+				return resp, err
+			}
+			if isMatch {
+				resp.MatchCount++
+			}
+		}
+	}
+	return resp, nil
+}
+
+// readFiles reads the content of files within the specified bucket with the
+// specified prefix path in parallel and returns their content. It fails if
+// operations to find or read any of the files fails.
+func readFiles(ctx context.Context, bucketName, prefix string) ([]string, error) {
+	type resp struct {
+		s   string
+		err error
+	}
+
+	client, err := storage.NewClient(ctx, option.WithoutAuthentication())
```

Comment:
Why without auth?

Reply:
Perhaps because it's a loopback (the server is spawned by the same process)?

Reply:
Marking this as resolved. Given that this is a loopback, I don't think we need to introduce the complexity of authentication.